stan/math/prim/fun/max_size.hpp

```diff
@@ -2,43 +2,21 @@
 #define STAN_MATH_PRIM_FUN_MAX_SIZE_HPP
 
 #include <stan/math/prim/fun/size.hpp>
+#include <algorithm>
 
 namespace stan {
 namespace math {
 
-template <typename T1, typename T2>
-size_t max_size(const T1& x1, const T2& x2) {
-  size_t result = stan::math::size(x1);
-  result = result > stan::math::size(x2) ? result : stan::math::size(x2);
-  return result;
-}
-
-template <typename T1, typename T2, typename T3>
-size_t max_size(const T1& x1, const T2& x2, const T3& x3) {
-  size_t result = stan::math::size(x1);
-  result = result > stan::math::size(x2) ? result : stan::math::size(x2);
-  result = result > stan::math::size(x3) ? result : stan::math::size(x3);
-  return result;
-}
-
-template <typename T1, typename T2, typename T3, typename T4>
-size_t max_size(const T1& x1, const T2& x2, const T3& x3, const T4& x4) {
-  size_t result = stan::math::size(x1);
-  result = result > stan::math::size(x2) ? result : stan::math::size(x2);
-  result = result > stan::math::size(x3) ? result : stan::math::size(x3);
-  result = result > stan::math::size(x4) ? result : stan::math::size(x4);
-  return result;
-}
-
-template <typename T1, typename T2, typename T3, typename T4, typename T5>
-size_t max_size(const T1& x1, const T2& x2, const T3& x3, const T4& x4,
-                const T5& x5) {
-  size_t result = stan::math::size(x1);
-  result = result > stan::math::size(x2) ? result : stan::math::size(x2);
-  result = result > stan::math::size(x3) ? result : stan::math::size(x3);
-  result = result > stan::math::size(x4) ? result : stan::math::size(x4);
-  result = result > stan::math::size(x5) ? result : stan::math::size(x5);
-  return result;
+/** Calculate the size of the largest input.
+ * @tparam T1 type of the first input
+ * @tparam Ts types of the other inputs
+ * @param x1 first input
+ * @param xs other inputs
+ * @return the size of the largest input
+ */
+template <typename T1, typename... Ts>
+inline size_t max_size(const T1& x1, const Ts&... xs) {
+  return std::max({stan::math::size(x1), stan::math::size(xs)...});
```

Comment:
Can this be generalized to handle size zero inputs as well as the existing generalization to one element?  Though I don't know what max returns with no inputs---negative infinity?  

```
template <typename... Ts>
 inline size_t max_size(const Ts&... xs) {
   return std::max({stan::math::size(xs)...});
```

Reply:
Yes. negative infinity isn't a size_t, but 0 is. I think where this comes up is if someone does max_size(xs...) and the parameter pack xs turns out to be empty, or maybe at the bottom of a meta programming recursion. I wonder what they would expect max_size([empty]) to be. A compiler error is a safer surprise than an unexpected 0, but 0 does seem like something someone might expect. Certainly no one would ever write deliberately write "max_size()".